plugins/modules/esxi_host.py

```diff
@@ -0,0 +1,406 @@
+        elif self.params['folder']:
+            if self.params['folder'].startswith(self.params['datacenter']) or self.params['folder'].startswith('/' + self.params['datacenter']):
+                path = self.params['folder']
+            else:
+                path = format_folder_path_as_host_fq_path(self.params['folder'], self.params['datacenter'])
+            self.folder = self.get_folder_by_absolute_path(folder_path=path, fail_on_missing=True)
+
+        self.host = self.get_esxi_host_by_name_or_moid(identifier=self.params['esxi_host_name'])
+
+    def __host_parent_type_is_folder(self):
+        """
+            Checks if the host is in a cluster or folder. Returns true if its in a folder.
+            Technically, the parent of a host in a folder is a ComputeResource. Clusters are
+            also a type of ComputeResource, so we need to check for the cluster type specifically.
+        """
+        if isinstance(self.host.parent, vim.ClusterComputeResource):
+            # the parent is a cluster
+            return False
+        else:
+            return True
+
+    def validate_maintenance_mode(func):   # pylint: disable=no-self-argument
+        """
+            Decorator function that adds a maintenance mode check before the wrapped method is called.
+        """
+        def wrapper(self):
+            if not self.host.runtime.inMaintenanceMode:
+                self.module.fail_json(msg='Host is not in maintenance mode. It must be in maintenance mode before it can be removed.')
+            func(self)
+        return wrapper
+
+    def create_host_connect_spec(self):
+        """
+            Function to create a host connection spec based on user params
+            Returns:
+                host connection specification
+        """
+        # Get the thumbprint of the SSL certificate
+        ssl_thumbprint = self.params['ssl_thumbprint']
+        if not ssl_thumbprint:
+            ssl_thumbprint = self.get_host_ssl_thumbprint()
+
+        host_connect_spec = vim.host.ConnectSpec()
+        host_connect_spec.sslThumbprint = ssl_thumbprint
+        host_connect_spec.hostName = self.params['esxi_host_name']
+        host_connect_spec.userName = self.params['esxi_username']
+        host_connect_spec.password = self.params['esxi_password']
+        host_connect_spec.force = self.params['force_add']
+        return host_connect_spec
+
+    def add_host(self):
+        host_connect_spec = self.create_host_connect_spec()
+        _kwargs = {'spec': host_connect_spec, 'license': None}
+        if self.folder:
+            task = self.folder.AddStandaloneHost(
+                **_kwargs, addConnected=self.params['add_connected'], compResSpec=None)
+        else:
+            task = self.cluster.AddHost_Task(
+                **_kwargs, asConnected=self.params['add_connected'], resourcePool=None)
+        try:
```

Comment:
the same `try/except` is repeated a few times. You can move it to separate `self.wait_for_task(task, error_msg)` method.